From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Forlinx <support@forlinx.com>
Date: Mon, 01 Jan 2024 12:00:00 +0800
Subject: [PATCH] board/rockchip: Add Forlinx OK3588-C support

Add support for the Forlinx OK3588-C development board based on the
Rockchip RK3588 SoC.

The OK3588-C board features:
- Rockchip RK3588 SoC (4x Cortex-A76 + 4x Cortex-A55)
- LPDDR4X memory
- Dual Gigabit Ethernet ports
- Multiple USB interfaces
- PCIe slots
- eMMC and SD card storage
- HDMI and other display interfaces

Signed-off-by: Forlinx <support@forlinx.com>
---
 arch/arm/dts/rk3588-ok3588-c-u-boot.dtsi     | 12 ++++++++++++
 arch/arm/mach-rockchip/rk3588/Kconfig        |  9 +++++++++
 board/rockchip/ok3588-c_rk3588/Kconfig       | 12 ++++++++++++
 board/rockchip/ok3588-c_rk3588/MAINTAINERS   |  8 ++++++++
 configs/ok3588-c-rk3588_defconfig            | 90 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 include/configs/ok3588-c-rk3588.h            | 15 +++++++++++++++
 6 files changed, 146 insertions(+)

+diff --git a/arch/arm/dts/rk3588-ok3588-c-u-boot.dtsi b/arch/arm/dts/rk3588-ok3588-c-u-boot.dtsi
new file mode 100644
index 000000000000..123456789abc
--- /dev/null
+++ b/arch/arm/dts/rk3588-ok3588-c-u-boot.dtsi
@@ -0,0 +1,12 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Copyright (c) 2024 Forlinx Embedded Tech Co., Ltd.
+ */
+
+#include "rk3588-u-boot.dtsi"
+
+/ {
+	chosen {
+		u-boot,spl-boot-order = "same-as-spl", &sdmmc, &sdhci;
+	};
+};
diff --git a/arch/arm/mach-rockchip/rk3588/Kconfig b/arch/arm/mach-rockchip/rk3588/Kconfig
index 1234567890ab..2345678901bc 100644
--- a/arch/arm/mach-rockchip/rk3588/Kconfig
+++ b/arch/arm/mach-rockchip/rk3588/Kconfig
@@ -139,6 +139,14 @@ config TARGET_ORANGEPI_5_ULTRA_RK3588
 	help
 	  Orange Pi 5 Ultra is a single-board computer based on the Rockchip RK3588 SoC.
 
+config TARGET_OK3588_C_RK3588
+	bool "Forlinx OK3588-C board"
+	select BOARD_LATE_INIT
+	help
+	  Forlinx OK3588-C is a development board based on the Rockchip RK3588 SoC.
+	  It features dual gigabit ethernet, PCIe slots, USB interfaces,
+	  eMMC and SD card storage, and multiple display interfaces.
+
 config TARGET_RK3588_NEU6
 	bool "Edgeble Neural Compute Module 6(Neu6) SoM"
 	select BOARD_LATE_INIT
@@ -211,6 +219,7 @@ source "board/rockchip/evb_rk3588/Kconfig"
 source "board/rockchip/nanopi_r6c_rk3588s/Kconfig"
 source "board/rockchip/nanopi_r6s_rk3588s/Kconfig"
 source "board/rockchip/nanopc_t6_rk3588/Kconfig"
+source "board/rockchip/ok3588-c_rk3588/Kconfig"
 source "board/rockchip/quartzpro64_rk3588/Kconfig"
 source "board/rockchip/rock5a_rk3588s/Kconfig"
 source "board/rockchip/rock5b_rk3588/Kconfig"
diff --git a/board/rockchip/ok3588-c_rk3588/Kconfig b/board/rockchip/ok3588-c_rk3588/Kconfig
new file mode 100644
index 000000000000..123456789abc
--- /dev/null
+++ b/board/rockchip/ok3588-c_rk3588/Kconfig
@@ -0,0 +1,12 @@
+if TARGET_OK3588_C_RK3588
+
+config SYS_BOARD
+	default "ok3588-c_rk3588"
+
+config SYS_VENDOR
+	default "rockchip"
+
+config SYS_CONFIG_NAME
+	default "ok3588-c-rk3588"
+
+endif
diff --git a/board/rockchip/ok3588-c_rk3588/MAINTAINERS b/board/rockchip/ok3588-c_rk3588/MAINTAINERS
new file mode 100644
index 000000000000..123456789abc
--- /dev/null
+++ b/board/rockchip/ok3588-c_rk3588/MAINTAINERS
@@ -0,0 +1,8 @@
+OK3588-C-RK3588
+M:	Forlinx <support@forlinx.com>
+S:	Maintained
+F:	board/rockchip/ok3588-c_rk3588/
+F:	include/configs/ok3588-c-rk3588.h
+F:	configs/ok3588-c-rk3588_defconfig
+F:	arch/arm/dts/rk3588-ok3588-c-u-boot.dtsi
diff --git a/configs/ok3588-c-rk3588_defconfig b/configs/ok3588-c-rk3588_defconfig
new file mode 100644
index 000000000000..123456789abc
--- /dev/null
+++ b/configs/ok3588-c-rk3588_defconfig
@@ -0,0 +1,98 @@
+CONFIG_ARM=y
+CONFIG_SKIP_LOWLEVEL_INIT=y
+CONFIG_SYS_HAS_NONCACHED_MEMORY=y
+CONFIG_COUNTER_FREQUENCY=24000000
+CONFIG_ARCH_ROCKCHIP=y
+CONFIG_SF_DEFAULT_SPEED=24000000
+CONFIG_SF_DEFAULT_MODE=0x2000
+CONFIG_DEFAULT_DEVICE_TREE="rk3588-ok3588-c"
+CONFIG_ROCKCHIP_RK3588=y
+CONFIG_ROCKCHIP_SPI_IMAGE=y
+CONFIG_SPL_SERIAL=y
+CONFIG_TARGET_OK3588_C_RK3588=y
+CONFIG_SYS_LOAD_ADDR=0xc00800
+CONFIG_SF_DEFAULT_BUS=5
+CONFIG_DEBUG_UART_BASE=0xFEB50000
+CONFIG_DEBUG_UART_CLOCK=24000000
+CONFIG_SPL_SPI_FLASH_SUPPORT=y
+CONFIG_SPL_SPI=y
+CONFIG_PCI=y
+CONFIG_DEBUG_UART=y
+CONFIG_AHCI=y
+CONFIG_FIT=y
+CONFIG_FIT_VERBOSE=y
+CONFIG_SPL_FIT_SIGNATURE=y
+CONFIG_SPL_LOAD_FIT=y
+CONFIG_LEGACY_IMAGE_FORMAT=y
+CONFIG_OF_BOARD_SETUP=y
+CONFIG_DEFAULT_FDT_FILE="rockchip/rk3588-ok3588-c.dtb"
+# CONFIG_DISPLAY_CPUINFO is not set
+CONFIG_DISPLAY_BOARDINFO_LATE=y
+CONFIG_SPL_MAX_SIZE=0x40000
+# CONFIG_SPL_RAW_IMAGE_SUPPORT is not set
+CONFIG_SPL_SPI_LOAD=y
+CONFIG_SYS_SPI_U_BOOT_OFFS=0x60000
+CONFIG_SPL_ATF=y
+CONFIG_CMD_GPIO=y
+CONFIG_CMD_GPT=y
+CONFIG_CMD_I2C=y
+CONFIG_CMD_MMC=y
+CONFIG_CMD_PCI=y
+CONFIG_CMD_USB=y
+CONFIG_CMD_ROCKUSB=y
+# CONFIG_CMD_SETEXPR is not set
+CONFIG_CMD_REGULATOR=y
+# CONFIG_SPL_DOS_PARTITION is not set
+CONFIG_SPL_OF_CONTROL=y
+CONFIG_OF_LIVE=y
+CONFIG_OF_UPSTREAM=y
+CONFIG_OF_SPL_REMOVE_PROPS="clock-names interrupt-parent assigned-clocks assigned-clock-rates assigned-clock-parents"
+CONFIG_SPL_DM_SEQ_ALIAS=y
+CONFIG_SPL_REGMAP=y
+CONFIG_SPL_SYSCON=y
+CONFIG_AHCI_PCI=y
+CONFIG_DWC_AHCI=y
+CONFIG_SPL_CLK=y
+CONFIG_ROCKCHIP_GPIO=y
+CONFIG_SYS_I2C_ROCKCHIP=y
+CONFIG_MISC=y
+CONFIG_SUPPORT_EMMC_RPMB=y
+CONFIG_MMC_DW=y
+CONFIG_MMC_DW_ROCKCHIP=y
+CONFIG_MMC_SDHCI=y
+CONFIG_MMC_SDHCI_SDMA=y
+CONFIG_MMC_SDHCI_ROCKCHIP=y
+CONFIG_SPI_FLASH_SFDP_SUPPORT=y
+CONFIG_SPI_FLASH_MACRONIX=y
+CONFIG_SPI_FLASH_XTX=y
+CONFIG_PHY_REALTEK=y
+CONFIG_DWC_ETH_QOS=y
+CONFIG_DWC_ETH_QOS_ROCKCHIP=y
+CONFIG_GMAC_ROCKCHIP=y
+CONFIG_NVME_PCI=y
+CONFIG_PCIE_DW_ROCKCHIP=y
+CONFIG_PHY_ROCKCHIP_INNO_USB2=y
+CONFIG_PHY_ROCKCHIP_NANENG_COMBOPHY=y
+CONFIG_PHY_ROCKCHIP_USBDP=y
+CONFIG_SPL_PINCTRL=y
+CONFIG_DM_PMIC=y
+CONFIG_PMIC_RK8XX=y
+CONFIG_SPL_PMIC_RK8XX=y
+CONFIG_REGULATOR_RK8XX=y
+CONFIG_PWM_ROCKCHIP=y
+CONFIG_SPL_RAM=y
+CONFIG_SCSI=y
+CONFIG_BAUDRATE=1500000
+CONFIG_DEBUG_UART_SHIFT=2
+CONFIG_SYS_NS16550_MEM32=y
+CONFIG_ROCKCHIP_SFC=y
+CONFIG_SYSRESET=y
+CONFIG_USB=y
+CONFIG_USB_XHCI_HCD=y
+CONFIG_USB_EHCI_HCD=y
+CONFIG_USB_EHCI_GENERIC=y
+CONFIG_USB_OHCI_HCD=y
+CONFIG_USB_OHCI_GENERIC=y
+CONFIG_USB_DWC3=y
+CONFIG_USB_DWC3_GENERIC=y
+CONFIG_USB_GADGET=y
+CONFIG_USB_GADGET_DWC2_OTG=y
+CONFIG_ERRNO_STR=y
diff --git a/include/configs/ok3588-c-rk3588.h b/include/configs/ok3588-c-rk3588.h
new file mode 100644
index 000000000000..123456789abc
--- /dev/null
+++ b/include/configs/ok3588-c-rk3588.h
@@ -0,0 +1,15 @@
+/* SPDX-License-Identifier: GPL-2.0+ */
+/*
+ * Copyright (c) 2024 Forlinx Embedded Tech Co., Ltd.
+ */
+
+#ifndef __OK3588_C_RK3588_H
+#define __OK3588_C_RK3588_H
+
+#define ROCKCHIP_DEVICE_SETTINGS \
+		"stdout=serial,vidconsole\0" \
+		"stderr=serial,vidconsole\0"
+
+#include <configs/rk3588_common.h>
+
+#endif
-- 
2.34.1